### ISSUE TITLE
Refactor the generation of code objects to get error messages early

### ISSUE THREAD

**Issue by mstimberg** [brian-team/brian2#405]
Most of the code generation machinery is run at the time of the run and this includes checking code for syntactic correctness, matching units, etc. This can lead to error messages at this late stage, with useless tracebacks (related issue: #376). Instead, we'd prefer to have errors as early as possible, whenever the incorrect code is written by the user, for example. In a few places, we've written explicit code to that effect (which also duplicates some work until #124 is implemented), but this is more of a hack, see this example from the `Thresholder` class:
```Python
        # Check the abstract code for unit mismatches (only works if the
        # namespace is already complete)
        try:
            self.update_abstract_code(level=1)
            check_code_units(self.abstract_code, self.group)
        except KeyError:
            pass
```

We should have a clean process consisting of multiple stages, where as much as possible is done as early as possible.

**Comment by thesamovar** [comment]
I'm actually wondering if it's worth considering a more substantial refactoring/rewrite of the codegen system. More and more it seems more complicated than it needs to be. Not sure if it's a good idea to be doing that again though.

**Comment by mstimberg** [comment]
I don't think we should do another really *substantial* refactoring, maybe except for the generator code (e.g. `make_statements` is much too long and complicated) -- this code is also quite independent from everything else. There are a few places that could be significantly simplified, though, I also feel we are passing around way too many arguments. Probably we should create an object (e.g. `CodeObject`) earlier in the process and store arguments in its attributes.

Maybe it would be a good idea to create an overview over what is currently happening (similar what you did for https://brian2.readthedocs.org/en/latest/developer/codegen.html#code-path) and see where things could be improved. For example, related to the problems I described in this issue, we should make a clearer distinction between code-generation-independent abstract preparations (variable name resolving, scalar and vector statements, ...) and the translation of code into a target language.

**Comment by thesamovar** [comment]
That sounds reasonable to me.

Yeah making an object we pass around rather than huge argument lists seems like a good idea. I thought that passing arguments was better at the beginning because I was hoping things could be modular. But it seems that actually it's very difficult to make them sufficiently modular so storing them in an object seems the better approach.

**Comment by thesamovar** [comment]
High priority for 2.0?

**Comment by thesamovar** [comment]
It seems unlikely we'll get this done in a reasonable time for 2.0, so let's just do #376 and postpone this.
